### PR TITLE
Feature/endpoint policy download

### PR DESCRIPTION
**Pull request by gaurav-gogia**
- add two new params `environment` and `access_token` for the config file
- download policies from authenticated endpoint
- convert policies into terrascan accepted format
- write test case for folder creation, file writing, file conversion conversion

### CODE REVIEW

File: pkg/initialize/run.go [hunk 1]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	if err != nil || res.StatusCode != 200 {
+		return fmt.Errorf("error downloading commercial policies. error: '%v', response status code: '%d'", err, res.StatusCode)
+	}
+	if res == nil {
+		return fmt.Errorf("error could not download policies, please check your network connection")
+	}
+
+	policies, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return fmt.Errorf("error reading api call response for commercial policies. error: '%v'", err)
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:120]: can we change this to `fmt.Errorf("error reading response body, error: %w, err")` ?
please where `%w` to wrap errors.

File: pkg/initialize/run.go [hunk 2]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+
+	err := json.Unmarshal(policies, &ruleMetadataList)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal policies into structure. error: '%v'", err)
+	}
+
+	for _, ruleMetadata := range ruleMetadataList {
+		policy, err := getCommercialPolicy(ruleMetadata)
+		if err != nil {
+			return err
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:137]: please wrap this error

File: pkg/initialize/run.go [hunk 3]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+		}
+
+		err = saveCommercialPolicies(policy, policyRepoPath)
+		if err != nil {
+			return err
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:142]: please wrap this error

File: pkg/initialize/run.go [hunk 4]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	policy.policyMetadata.Severity = ruleMetadata.Severity
+	policy.policyMetadata.Description = ruleMetadata.RuleDisplayName
+	policy.policyMetadata.ReferenceID = ruleMetadata.RuleReferenceID
+	policy.policyMetadata.ID = ruleMetadata.RuleReferenceID
+	policy.policyMetadata.Category = ruleMetadata.Category
+	policy.policyMetadata.Version = ruleMetadata.Version
+
+	templateString, ok := ruleMetadata.RuleArgument.(string)
+	if !ok && templateString != "" {
+		return policy, fmt.Errorf("incorrect rule argument type, must be a []byte")
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:169]: you are asserting that RuleArgument is a string and logging that it must be []byte, is that correct?

File: pkg/initialize/run.go [hunk 5]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	policy.policyMetadata.ResourceType = ruleMetadata.ResourceType
+	policy.policyMetadata.Severity = ruleMetadata.Severity
+	policy.policyMetadata.Description = ruleMetadata.RuleDisplayName
+	policy.policyMetadata.ReferenceID = ruleMetadata.RuleReferenceID
+	policy.policyMetadata.ID = ruleMetadata.RuleReferenceID
+	policy.policyMetadata.Category = ruleMetadata.Category
+	policy.policyMetadata.Version = ruleMetadata.Version
+
+	templateString, ok := ruleMetadata.RuleArgument.(string)
+	if !ok && templateString != "" {
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:168]: if the assertion fails, do we need to the second check?

File: pkg/initialize/run.go [hunk 6]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+		return policy, fmt.Errorf("incorrect rule argument type, must be a []byte")
+	}
+	err := json.Unmarshal([]byte(templateString), &templateArgs)
+	if err != nil {
+		return policy, fmt.Errorf("error could not unmarshal rule arguments into map[string]interface{}: '%v'", err)
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:173]: can we change the message to `error occurred while unmarshaling rule arguments`.
please use `%w` for wrapping errors in all wrapped errors.

File: pkg/initialize/run.go [hunk 7]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	regoPath := filepath.Join(resourceDir, policy.policyMetadata.File)
+	err = ioutil.WriteFile(regoPath, []byte(policy.regoTemplate), filePermissionBits)
+	if err != nil {
+		return fmt.Errorf("error could not write rego code file on disk: '%v'", err)
+	}
+
+	return nil
+}
+
+func getCSP(resourceType string) string {
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:222]: please change the function name, `getCSP` is unclear.

File: pkg/initialize/run.go [hunk 8]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+
+	return strings.Split(csp, "_")[0]
+}
+
+func ensureDir(path string) error {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		err = os.Mkdir(path, filePermissionBits)
+		if err != nil {
+			return fmt.Errorf("error unable to create requested directory: '%v'", err)
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:245]: we should also log the `path`.
please change the error message.

File: pkg/initialize/run_test.go [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+const TESTDIR = "./test_data/"
```
Comment (patilpankaj212) [inline pkg/initialize/run_test.go:30]: it can be "test_data"

File: pkg/initialize/run_test.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+func TestGetCommercialPolicy(t *testing.T) {
+	policies, err := ioutil.ReadFile("./test_data/policies.json")
```
Comment (patilpankaj212) [inline pkg/initialize/run_test.go:33]: please use `filepath.Join()`

File: pkg/initialize/run_test.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+)
+
+const TESTDIR = "./test_data/"
+
+func TestGetCommercialPolicy(t *testing.T) {
```
Comment (patilpankaj212) [inline pkg/initialize/run_test.go:32]: can we add more tests for error scenarios as well? please see tests in other packages.

File: pkg/initialize/types.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
+package initialize
```
Comment (patilpankaj212) [inline pkg/initialize/types.go:1]: file headers

File: pkg/initialize/run.go [hunk 13]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	if err != nil {
+		return policy, fmt.Errorf("error could not unmarshal rule arguments into map[string]interface{}: '%v'", err)
+	}
+	policy.policyMetadata.TemplateArgs = templateArgs
+
+	return policy, nil
+}
+
+func saveCommercialPolicies(policy commercialPolicy, policyRepoPath string) error {
+	const tabSpace = "    "
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:181]: this is only used once

File: pkg/initialize/run.go [hunk 14]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	if err != nil {
+		return fmt.Errorf("error could not marshal json object into byte array: '%v'", err)
+	}
+
+	metadata := buffer.Bytes()
+	metadata = bytes.TrimRight(metadata, "\n")
+	metaDataPath := filepath.Join(resourceDir, policy.metadataFileName)
+	err = ioutil.WriteFile(metaDataPath, metadata, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("error could not write rule metadata file on disk: '%v'", err)
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:210]: we can use following error messages
`fmt.Errorf("error occurred while writing rule metedata file, %w", err)`
or 
`fmt.Errorf("could not write rule metedata file, error: %w", err)`

File: pkg/initialize/run.go [hunk 15]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	}
+
+	regoPath := filepath.Join(resourceDir, policy.policyMetadata.File)
+	err = ioutil.WriteFile(regoPath, []byte(policy.regoTemplate), filePermissionBits)
+	if err != nil {
+		return fmt.Errorf("error could not write rego code file on disk: '%v'", err)
```
Comment (patilpankaj212) [inline pkg/initialize/run.go:216]: we can use following error messages
`fmt.Errorf("error occurred while writing rego file, %w", err)`
or
`fmt.Errorf("could not write rego file, error: %w", err)`

**Review by patilpankaj212 — CHANGES_REQUESTED** [review]

File: pkg/initialize/run.go [hunk 16]
```diff
@@ -60,20 +63,202 @@ func Run(isNonInitCmd bool) error {
...
+	if err != nil {
+		return fmt.Errorf("error constructing request object. error: '%v'", err)
+	}
+	req.Header.Add("Authorization", "Bearer "+accessToken)
+
+	res, err := client.Do(req)
+	if err != nil || res.StatusCode != 200 {
+		return fmt.Errorf("error downloading commercial policies. error: '%v', response status code: '%d'", err, res.StatusCode)
+	}
+	if res == nil {
```
[the remark this answers is not in the document]
Comment (gaurav-gogia) [reply pkg/initialize/run.go:114]: yes, when my wifi wasn't connected and i didn't notice that it wasn't

**Review by cesar-rodriguez — COMMENTED** [review]
Can you please also update the docs here: https://github.com/accurics/runterrascan.io?

**Review by patilpankaj212 — APPROVED** [review]
LGTM !!

**Pull request merged**